sdk/go/common/apitype/search.go

```diff
@@ -0,0 +1,120 @@
+	Pagination   *ResourceSearchPagination `json:"pagination,omitempty"`
+}
+
+func (r *ResourceSearchResponse) Error() error {
+	return fmt.Errorf("just a normal error")
+}
+
+// ResourceResult is the user-facing type for our indexed resources.
+//
+// If you add a property here, don't forget to update fieldMappings to make it
+// queryable!
+type ResourceResult struct {
+	Created      *string                      `json:"created,omitempty" csv:"created"`
+	Custom       *bool                        `json:"custom,omitempty" csv:"custom"`
+	Delete       *bool                        `json:"delete,omitempty" csv:"delete"`
+	Dependencies []string                     `json:"dependencies,omitempty" csv:"-"`
+	ID           *string                      `json:"id,omitempty" csv:"id"`
+	Matches      map[string][]json.RawMessage `json:"matches,omitempty" csv:"-"`
+	Modified     *string                      `json:"modified,omitempty" csv:"modified"`
+	Module       *string                      `json:"module" csv:"module"`
+	Name         *string                      `json:"name,omitempty" csv:"name"`
+	Package      *string                      `json:"package" csv:"package"`
+	Parent       *string                      `json:"parent.urn,omitempty" csv:"parent_urn"`
+	Pending      *string                      `json:"pending,omitempty" csv:"pending"`
+	Program      *string                      `json:"project,omitempty" csv:"project"`
+	Protected    *bool                        `json:"protected,omitempty" csv:"protected"`
+	ProviderURN  *string                      `json:"provider.urn,omitempty" csv:"provider_urn"`
+	Stack        *string                      `json:"stack,omitempty" csv:"stack"`
+	Type         *string                      `json:"type,omitempty" csv:"type"`
+	URN          *string                      `json:"urn,omitempty" csv:"urn"`
+	Teams        teamNameSlice                `json:"teams,omitempty" csv:"teams"`
+	Properties   *RawProperty                 `json:"properties,omitempty" csv:"properties"`
+}
+
+// Aggregation collects the top 5 aggregated values for the requested dimension.
+type Aggregation struct {
+	Others  *int64              `json:"others,omitempty"`
+	Results []AggregationBucket `json:"results,omitempty"`
+}
+
+// AggregationBucket is an example value for the requested aggregation, with a
+// count of how many resources share that value.
+type AggregationBucket struct {
+	Name  *string `json:"name,omitempty"`
+	Count *int64  `json:"count,omitempty"`
+}
+
+// ResourceSearchPagination provides links for paging through results.
+type ResourceSearchPagination struct {
+	Previous *string `json:"previous,omitempty"`
+	Next     *string `json:"next,omitempty"`
+	Cursor   *string `json:"cursor,omitempty"`
+}
+
+type teamNameSlice []string
+
+func (teams *teamNameSlice) MarshalCSV() (string, error) {
+	if teams == nil || len(*teams) == 0 {
+		return "", nil
+	}
+
+	json, err := json.Marshal(*teams)
```

Comment:
Should it be `json.Marshal` even if `*teams` is empty? `[]`?

Reply:
Sorry, let me clarify. The condition above checks if `*teams` is nil, and if yes, it returns an empty string. That's fine, in that if you try to serialize a pointer to `*teams`, it should just 🤷 

However, if `*teams` is non-nil, you have a possibly-empty slice.
Right now, it seems that we'll serialize `["a", "b"]` to CSV if the slice is non-empty, and an empty string (``) if the slice is empty—versus an empty list `[]`. 
I'm suggesting that we might want more consistent output for the CSV? "The 'teams' field is always a JSON array" is better than "it's a JSON array when non-empty, and blank otherwise".

Reply:
Aha! That makes a TON of sense. Thank you for the clarification (and agreed - better to always return the same type where possible)

Reply:
Yeah, I wouldn't marshal `[]string{}` at all. Just `return "[]"`  is a valid empty JSON array if that's what you want. But I was imagining:

```go
func (teams *teamNameSlice) MarshalCSV() (string, error) {
  if *teams == nil { return "", nil }  // (1)
  
  bs, err := json.Marshal(*teams)      // (2)
  return string(bs), err
}
```

1. `teamNameSlice` should not be used as a pointer usually since it'll normally be used as the following, so the pointer in MarshalCSV will always be non-nil.

    ```go
    type Foo struct { Teams teamNameSlice }
    ```

    However, if it's used as a pointer, and the pointer is nil:

    ```go
    type Foo struct { Teams *teamNameSlice }

    f := Foo{Teams: nil}
    ```

    Then the user probably doesn't want us to write anything for this field, so it makes sense to return an empty string.
    
2. Otherwise, marshal as-is and return it. If the slice is empty, it'll get `[]`, and otherwise it'll get `["foo", "bar"]`.

Reply:
That makes perfect sense, and is much cleaner than what I had. Thanks again!